### ISSUE TITLE
[Feature Request] Add the option to add/create rule based on Template to JSRule or Rule Builder

### ISSUE THREAD

**Issue by ThaDaVos** [openhab/openhab-js#363]
It would be great if one could instantiate a rule template from a script in the `automations/js` folder - this way one can create the necessary items/groups etc and instantiate the rule template from one place - so when the script is removed, all gets cleaned up. (At least this is my understanding how `automation/js` is used.

## Your Environment

* Version used: (e.g., openHAB and **JS Scripting add-on version**) 4.3.0 **4.3.0.SNAPSHOT**
* Environment name and version (e.g. Chrome 76, Java 8, Node.js 12.9, ...): Docker Container
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...): Docker Container

**Comment by florian-h05** [comment]
JSRule and Rule Builder work different to UI-based rules wrt to their structure and rule templates are a UI-only feature.

It is possible to create something like JSRule Templates, e.g. by writing a function that creates the rule (I do this really much, e.g. https://github.com/florian-h05/openhab-js-tools/blob/f5e48a2ec0325df920c4ba15dfd1a3d77973220b/src/thingsx/health.js#L129, but instantiating UI-rule templates from script languages is not possible I think and would be high effort to get it working. 

I think this feature request is a "Won't fix" and I will therefore close this, sorry.

**Comment by ThaDaVos** [comment]
It's sad that this is marked as `Won't fix` as it would would mean one has to copy the templates manually from marketplace to include them when all other rules are file-based - I rather have my rules file-based so I can version control them

**Comment by florian-h05** [comment]
I think it is easier to rewrite this template to a rule generator function in that case — as long as one is proficient enough with writing code.
I can re-open this feature request, but I personally won‘t work on it and as long as no one else contributes this, I guess the feature request will not be implemented.

**Comment by florian-h05** [comment]
To implement this, one would need to check how rule templates are handled in detail, if it is UI-only or if openHAB core does the work of applying the template (i.e. add triggers, actions, conditions of the template to the rule). If is done in core, it should be easier to do that from scripts.

**Comment by ThaDaVos** [comment]
According to below file it's added to the persistent storage:
[MarketplaceRuleTemplateProvider.java](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.addon.marketplace/src/main/java/org/openhab/core/addon/marketplace/internal/automation/MarketplaceRuleTemplateProvider.java)

And it seems it's also stored inside the Rule itself:
[Rule.java#L67](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/Rule.java#L67)

And there's also a constructor for it:
[RuleImpl.java#L90](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/RuleImpl.java#L90)

So with a quick search it seems it's all done inside the `openhab-core` luckily - I haven't checked how it's wrapped from `openhab-js` side yet, will do that when I have time

**Comment by ThaDaVos** [comment]
Could it be as simple as allowing it to be passed here?
[rules.js#L258](https://github.com/openhab/openhab-js/blob/main/src/rules/rules.js#L258)

**Comment by florian-h05** [comment]
Probably yes - we need to check if it is working and if lifecycle is properly handled. But this should be an easy test.

**Comment by ThaDaVos** [comment]
Yeah I was thinking about manually editing the source code by manually loading it inside Openhab (through the manual install option) and seeing what it does - probably won't be today - think weekend on it's earliest

**Comment by florian-h05** [comment]
I don't think it will be as easy as we hoped - file-based/unmanaged rules created through scripts are managed by the [`ScriptedAutomationManager`](https://github.com/openhab/openhab-core/blob/f502e9d4e3ad8f1cc9521de02da097725644f0c5/bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/shared/ScriptedAutomationManager.java#L48), which handles the rules different from UI-based rules.

**Comment by ThaDaVos** [comment]
Could be easier then you think - seeing the following line:
[RuleBuilder.java#L60](https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/util/RuleBuilder.java#L60)

**Comment by ThaDaVos** [comment]
So something as follows is needed to pass it through I think:
https://github.com/ThaDaVos/openhab-core/commit/027958b397266e1b73d0d8ac6575ae616394734e

**Comment by florian-h05** [comment]
> I added the check if ruleUid != templateUid because of [rules.js#L258](https://github.com/openhab/openhab-js/blob/main/src/rules/rules.js#L258)

Rather fix this line in rules.js.

**Comment by rkoshak** [comment]
See https://github.com/openhab/openhab-core/pull/4633 which sadly didn't make it into 5.0 release but hopefully will be coming along soon. 

Because rule templates are not a JS specific feature, this shouldn't be implemented here. 

> It's sad that this is marked as Won't fix as it would would mean one has to copy the templates manually from marketplace to include them when all other rules are file-based 

Rule templates, as currently implemented are *installed* like add-ons, not copy and pasted. That's why only templates are supported only for managed rules. 

However, with the new additions in OH 5 and the YAML file format, once the PR above is merged you will be able to install a template and instantiate a rule based on that template from a file, not just the UI. I believe it also means you'll be able write rule templates locally so you don't need to install them from the marketplace. 

I recommend closing this issue as it's something that should be implemented in core in the first place and a PR to implement it in core is awaiting review.
